cmd/kubeadm/app/phases/controlplane/components.go

```diff
@@ -0,0 +1,183 @@
+
+	client, err := rest.HTTPClientFor(&rest.Config{
+		TLSClientConfig: rest.TLSClientConfig{
+			CertFile: certFile,
+			KeyFile:  keyFile,
+			Insecure: true,
+		},
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to create kubelet client")
+	}
+
+	kubeletEndpoint, err := getKubeletEndpoint(filepath.Join(kubeletDir, kubeadmconstants.KubeletConfigurationFileName))
+	if err != nil {
+		return errors.Wrap(err, "failed to get kubelet endpoint")
+	}
+
+	components := make([]*component, len(componentNames))
+	for i, name := range componentNames {
+		labels, err := getComponentLabels(name, manifestDir)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get pod labels of %s component", name)
+		}
+
+		components[i] = &component{name, labels, false}
+	}
+
+	return wait.PollUntilContextTimeout(context.Background(), 5*time.Second, timeout, false, func(ctx context.Context) (bool, error) {
+		klog.V(1).Infoln("[control-plane] polling status of control plane components...")
+
+		resp, err := client.Get(kubeletEndpoint)
+		if err != nil {
+			fmt.Printf("[kubelet client] Error getting pods [%v]\n", err)
+			return false, nil
+		}
+
+		defer resp.Body.Close()
+
+		data, err := io.ReadAll(resp.Body)
+		if err != nil {
+			fmt.Printf("[kubelet client] Error reading pods from response body [%v]\n", err)
+			return false, nil
+		}
+
+		pods := &v1.PodList{}
+		if err := json.Unmarshal(data, pods); err != nil {
+			fmt.Printf("[kubelet client] Error parsing pods from response body: %q\n", data)
+			return false, nil
+		}
+
+		for _, comp := range components {
+			labels := comp.labels
+		match_pod:
+			for _, pod := range pods.Items {
+				podLabels := pod.ObjectMeta.Labels
+				for key, value := range labels {
+					if podLabels[key] != value {
+						continue match_pod
+					}
+				}
+
+				comp.touched = true
+
+				for _, status := range pod.Status.ContainerStatuses {
+					if !status.Ready {
+						klog.V(1).Infof("[control-plane] component: %s is not ready\n", comp.name)
+						return false, nil
+					}
+				}
+
+				klog.V(1).Infof("[control-plane] component: %s is ready\n", comp.name)
+			}
+		}
+
+		for _, comp := range components {
+			if !comp.touched {
+				fmt.Printf("[kubelet client] Couldn`t find pod for component: %s with labels: [%v]\n", comp.name, comp.labels)
+				return false, nil
+			}
+		}
+
+		return true, nil
+	})
+}
+
+func getKubeletEndpoint(configFile string) (string, error) {
+	config := &kubeletconfig.KubeletConfiguration{}
+
+	data, err := os.ReadFile(configFile)
+	if err != nil {
+		return "", err
+	}
+
+	if err := yaml.Unmarshal(data, config); err != nil {
+		return "", err
+	}
+
+	if config.Authorization.Mode == kubeletconfig.KubeletAuthorizationModeWebhook {
+		// make sure cluster admins role binding is created, thus request to kubelet will pass server authorization
+		if _, err := kubeconfig.EnsureAdminClusterRoleBinding(kubeadmconstants.KubernetesDir, nil); err != nil {
```

Comment:
isn't the RB already created at this point?
it is one of the first things that happen during kubeadm init and for joining CP nodes it should be there already.

Reply:
> isn't the RB already created at this point? it is one of the first things that happen during kubeadm init and for joining CP nodes it should be there already.

This RB is created in function `initData.Client()` during kubeadm init, and `initData.Client()` is ever not called till `WaitControlPlanePhase` is reached

Reply:
if the CRB is required earlier perhaps we can call a `_ = initData.Client()` instead of calling `kubeconfig.EnsureAdminClusterRoleBinding`. that is because we want this Client() function to set the right vars / state. getKubeletEndpoint() should just return the endpoint and not perform CRB creation.

in runWaitControlPlanePhase there is a `client, err := data.ClientWithoutBootstrap()`, but this client doesn't have permissions.

in this discussion we talked about using the apiserver client cert/key
https://github.com/kubernetes/kubernetes/pull/119598/files#r1430504568
and the code in this PR is still using it. i'm not sure i understand why we need to bootstrap the admin.conf client?

i.e.
> // make sure cluster admins role binding is created, thus request to kubelet will pass server authorization

Reply:
> but, in this discussion we talked about using the apiserver client cert/key
> https://github.com/kubernetes/kubernetes/pull/119598/files#r1430504568
> and the code in this PR is still using it. i'm not sure i understand why we need to bootstrap the admin.conf client?

previously `waitconrolpane,go` use `data.Client()` to get client which is changed by this [commit](https://github.com/kubernetes/kubernetes/commit/6dc11c1f219a53f4f9cd28f44a569b19927004b1).

> kubeconfig.EnsureAdminClusterRoleBinding() ensures that the kubeadm:cluster-admins group (which admin.conf is part of) is grated the cluster-admin privilege. it seems unrelated to getting the pods with the apiserver client cert/key. that user should already be able to get the pods from the kubelet, no?

kubelet server will [install auth filter](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/auth.go#L116) with `Authorization` configuration, and when `Authorization.Mode = Webhook` (which is default),  kubelet server will create `SubjectAccessReview`  to authorize requests to kubelet. 

> if the CRB is required earlier perhaps we can call a _ = initData.Client() instead of calling kubeconfig.EnsureAdminClusterRoleBinding. that is because we want this Client() function to set the right vars / state. getKubeletEndpoint() should just return the endpoint and not perform CRB creation.

Why put this code in `getKubeletEndpoint()` is only use `_ = initData.Client()` is not clear why request to kubelet will be failed without `kubeadm:cluster-admins` RB is created (I did spend some time to find why request to kubelet return athorize error). 

Now, I need to revert `client, err := data.ClientWithoutBootstrap()` to `client, err := data.Client()` or `_ = data.Client()` when this feature gate is enabled in waitcontrolplane?

Reply:
> Now, I need to revert client, err := data.ClientWithoutBootstrap() to client, err := data.Client() or _ = data.Client() when this feature gate is enabled in waitcontrolplane?

there might be a "chicken and egg" problem, because  data.ClientWithoutBootstrap was used because  data.Client() does not work yet. i.e. the CRB cannot be performed due to apiserver pod not ready. you can try it, of course.

> kubelet server will [install auth filter](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/auth.go#L116) with Authorization configuration, and when Authorization.Mode = Webhook (which is default), kubelet server will create SubjectAccessReview to authorize requests to kubelet.

i see, so it's because of this:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/certs/certlist.go#L294
apiserver-kubelet-client needs the admin credentials, to get pods or do other actions.

if it's possible to create the Client() earlier that might be best, but if it's not, perhaps we can:
- during init use the credentials from super-admin.conf
- during join just use admin.conf or apiserver-kubelet-client. at that point the CRB should already exist.